crates/accelerate/src/twirling.rs

```diff
@@ -0,0 +1,181 @@
+    ([XGate, ZGate, YGate, YGate], PI),
+    ([XGate, XGate, XGate, IGate], 0.),
+    ([YGate, IGate, YGate, XGate], 0.),
+    ([YGate, ZGate, XGate, YGate], 0.),
+    ([YGate, XGate, YGate, IGate], 0.),
+    ([YGate, YGate, XGate, ZGate], PI),
+];
+
+static CZ_TWIRL_SET: [([StandardGate; 4], f64); 16] = [
+    ([IGate, ZGate, IGate, ZGate], 0.),
+    ([IGate, XGate, ZGate, XGate], 0.),
+    ([IGate, YGate, ZGate, YGate], 0.),
+    ([IGate, IGate, IGate, IGate], 0.),
+    ([ZGate, XGate, IGate, XGate], 0.),
+    ([ZGate, YGate, IGate, YGate], 0.),
+    ([ZGate, IGate, ZGate, IGate], 0.),
+    ([ZGate, ZGate, ZGate, ZGate], 0.),
+    ([XGate, YGate, YGate, XGate], PI),
+    ([XGate, IGate, XGate, ZGate], 0.),
+    ([XGate, ZGate, XGate, IGate], 0.),
+    ([XGate, XGate, YGate, YGate], 0.),
+    ([YGate, IGate, YGate, ZGate], 0.),
+    ([YGate, ZGate, YGate, IGate], 0.),
+    ([YGate, XGate, XGate, YGate], PI),
+    ([YGate, YGate, XGate, XGate], 0.),
+];
+
+static ISWAP_TWIRL_SET: [([StandardGate; 4], f64); 16] = [
+    ([IGate, ZGate, ZGate, IGate], 0.),
+    ([IGate, XGate, YGate, ZGate], 0.),
+    ([IGate, YGate, XGate, ZGate], PI),
+    ([IGate, IGate, IGate, IGate], 0.),
+    ([ZGate, XGate, YGate, IGate], 0.),
+    ([ZGate, YGate, XGate, IGate], PI),
+    ([ZGate, IGate, IGate, ZGate], 0.),
+    ([ZGate, ZGate, ZGate, ZGate], 0.),
+    ([XGate, YGate, YGate, XGate], 0.),
+    ([XGate, IGate, ZGate, YGate], 0.),
+    ([XGate, ZGate, IGate, YGate], 0.),
+    ([XGate, XGate, XGate, XGate], 0.),
+    ([YGate, IGate, ZGate, XGate], PI),
+    ([YGate, ZGate, IGate, XGate], PI),
+    ([YGate, XGate, XGate, YGate], 0.),
+    ([YGate, YGate, YGate, YGate], 0.),
+];
+
+#[pyfunction]
+#[pyo3(signature=(circ, twirled_gate, seed=None, num_twirls=1))]
+pub fn twirl_circuit(
+    py: Python,
+    circ: &CircuitData,
+    twirled_gate: StandardGate,
+    seed: Option<u64>,
+    num_twirls: usize,
+) -> PyResult<Vec<CircuitData>> {
+    let mut rng = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+    let twirl_set: &[([StandardGate; 4], f64)] = match twirled_gate {
+        StandardGate::CXGate => &CX_TWIRL_SET,
+        StandardGate::CZGate => &CZ_TWIRL_SET,
+        StandardGate::ECRGate => &ECR_TWIRL_SET,
+        StandardGate::ISwapGate => &ISWAP_TWIRL_SET,
+        _ => {
+            return Err(QiskitError::new_err(
+                "Provided gate to twirl is not currently supported you can only use CX, CZ, ECR or iSwap.",
+            ))
+        }
+    };
+    let generate_twirled_circuit = |rng: &mut Pcg64Mcg| {
+        let mut out_circ = CircuitData::clone_empty_from(circ, None);
+
+        for inst in circ.data() {
+            match inst.op.view() {
+                OperationRef::Standard(gate) => {
+                    if gate == twirled_gate {
+                        let qubits: Vec<Qubit> = out_circ.get_qargs(inst.qubits).to_vec();
```

Comment:
This was necessary because the output references from `get_qargs` are from `out_circ` and we mutate `out_circ` below so the borrow checker wasn't happy because we have immutable references here when trying to get a mutable reference for the push calls.

Reply:
Ah sorry, missed that.  Since we're already making the assumption that the `qargs` interners are equivalent to the base circuit (from `clone_empty_like` and the subsequent direct `push(inst.clone()`), can we instead use the input circuit to get the qargs, and avoid the vec allocation?

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/13331/commits/f70566df24e7e9f51a0e6aafe04d48f4a00ab4d6